implement dashboards

stephenplusplus opened the pull request:
### Demo

http://stephenplusplus.github.io/phoenix/#/projects

<img width="749" alt="screen shot 2015-11-04 at 2 46 14 pm" src="https://cloud.githubusercontent.com/assets/723048/10949824/d910f9c8-8302-11e5-8128-cd5f14e64e99.png">

<img width="839" alt="screen shot 2015-11-05 at 9 57 29 am" src="https://cloud.githubusercontent.com/assets/723048/10971713/a9cfabd6-83a3-11e5-955a-1668eabcdd43.png">

<img width="982" alt="screen shot 2015-11-05 at 11 40 54 am" src="https://cloud.githubusercontent.com/assets/723048/10974890/1da6a63c-83b2-11e5-9d7e-91a5d049ddac.png">
#### To Dos

**Display all dashboards for a project**
- [x] All dashboards created for a project will be visible to any user who has permissions to view that project
- [x] Separate a user's own dashboards from the project's dashboards that they didn't create

**Allow using Popular Dashboards**
- [x] Create a whitelist of the official dashboards:
  - Data Analysis: `data-analysis`
  - Media Management: `media-management`
  - Web Hosting: `web-hosting`
- [x] Display the whitelist after selecting the "+" on the tab that lists a project's dashboards

**Allow any dashboard to be shared**
- [x] Selecting the share icon on a dashboard will open a dialog which gives the user a unique ID like those above. If there's a conflict, append `-Date.now()` after the slug
- [x] Design an "Import" tab in the "Create Dashboard" dialog that accepts a dashboard ID (from the above step). It will:
  - Display the name of the plugin (allow it to be changed)
  - The names of the plug-ins it uses
- [ ] Filter out private data before publishing

**Support configuring a dashboard**
- [ ] editing its name
- [x] deleting it
- [ ] un-doing the delete

**Odds and ends**
- [ ] Redirect `/projects/{projectId}/` to homepage view with the project's tab selected

stephenplusplus committed:
implement dashboards

callmehiphop commented:
:+1: I should be good to meet up any day this week

stephenplusplus commented:
// @jgeewax dashboard preview

callmehiphop commented:
So I think the easiest route is to just follow material design [card guidelines](https://www.google.com/design/spec/components/cards.html#cards-content). I think our cards would resemble the Google Notes example, where we use a solid card with a divider for our actions. 

I cooked up an example of what I'm thinking, it's missing a few things, but I think you'll get the gist.

<img width="1006" alt="screen shot 2015-11-03 at 2 07 19 pm" src="https://cloud.githubusercontent.com/assets/1707267/10918592/814eadd0-8234-11e5-822b-fdfe1df9879d.png">

And for mobile view we'd just go into a column mode?

<img width="398" alt="screen shot 2015-11-03 at 2 02 53 pm" src="https://cloud.githubusercontent.com/assets/1707267/10918601/8f719094-8234-11e5-96c1-0cc52ea884e1.png">

jgeewax commented:
Lgtm

stephenplusplus commented:
Lgtmt. Well done sir!

On Tuesday, November 3, 2015, JJ Geewax <notifications@github.com> wrote:

> Lgtm
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jgeewax/phoenix/pull/45#issuecomment-153458045>.
>

callmehiphop commented:
:dancer:

callmehiphop commented:
What are our feelings on using [subheaders](https://material.angularjs.org/latest/demo/subheader) to separate between popular/project/etc. dashboards? I think from a desktop perspective it probably won't matter, but I could see it being nice for mobile.

stephenplusplus commented:
Give it a shot? I think I tried it originally, but couldn't come up with a combination of headings and styles that made each piece feel like it was visually accomplishing the separation.

stephenplusplus commented:
@callmehiphop is applying the new, awesome styles to the branch now.

To be completed:

- [ ] All dashboards created for a project will be visible to any user who has permissions to view that project
  - :question: Should only the creator of a dashboard be able to edit the plugins/configuration, or is being a part of the project be good enough?
  - These dashboards are separated from the dashboards a user has created themselves
- [ ] Create dummy dashboards, displayed in two places:
  - After selecting the "+" on the tab that lists a project's dashboards
  - Within their own tab, "Popular Dashboards"
    - This will be a curated list, to start:
      - **Media Management** -- a dashboard, including these plugins:
        - Media Metadata
        - Render Watcher
      - **Web Hosting**
        - Storage Browser
        - Cloud SQL Viewer
        - VM Monitor
        - Pub/Sub Tracer
      - **Data Analysis**
        - BigQuery Visualizer
- [ ] Allow any dashboard to be shared
  - Selecting the share icon on a dashboard will open a dialog which gives the user a URL, like: "/#/import/{{id}}". That will:
    - Have a dialog open showing:
      - the name of the dashboard
      - the plugins it has installed
      - a dropdown that allows the user to select which project to apply it to

I think that's it, lmk if I missed something.

stephenplusplus commented:
New previews!

![](https://cloud.githubusercontent.com/assets/1707267/10945093/4c5ae45c-82ea-11e5-9076-9ab39e95c662.png)

![](https://cloud.githubusercontent.com/assets/1707267/10945102/54ff2460-82ea-11e5-9c03-88486f4789a5.png)

stephenplusplus commented:
TODO: separate private data from non-private when shared

stephenplusplus commented:
Added more todos to the first post

stephenplusplus commented:
It looks so prett-ay. Deploying to my gh-pages now.

<img width="407" alt="screen shot 2015-11-04 at 9 20 22 pm" src="https://cloud.githubusercontent.com/assets/723048/10958225/194f249c-833a-11e5-8dc1-e2a01224cb23.png">

This kinda looks like an input I'm supposed to type in. Can it be one of those "clicking in this window selects all of the text" kind of things?

--

<img width="400" alt="screen shot 2015-11-04 at 9 23 04 pm" src="https://cloud.githubusercontent.com/assets/723048/10958253/48f55158-833a-11e5-90fe-4e8729a080fe.png">

The error text isn't as error-y anymore. Previously, it looked like this:

<img width="400" src="https://cloud.githubusercontent.com/assets/723048/10949824/d910f9c8-8302-11e5-8128-cd5f14e64e99.png">

-- 

Flipping over to "Import", it still shows the error message:

<img width="405" alt="screen shot 2015-11-04 at 9 24 12 pm" src="https://cloud.githubusercontent.com/assets/723048/10958275/6f935044-833a-11e5-9172-2428b3edde0a.png">

--

Then, when I try to type in a code, I get an error in the console:

I'm getting a bug when I type in a code:

```
TypeError: timeoutHandle.cancel is not a function
    at Object.fn (projects.controller.js:49)
    at Scope.$digest (angular.js:15826)
    at Scope.$apply (angular.js:16097)
    at $$debounceViewValueCommit (angular.js:25388)
    at $setViewValue (angular.js:25360)
    at HTMLInputElement.listener (angular.js:21679)
    at HTMLInputElement.eventHandler (angular.js:3298)(anonymous function) @ angular.js:12477(anonymous function) @ angular.js:9246Scope.$digest @ angular.js:15844Scope.$apply @ angular.js:16097$$debounceViewValueCommit @ angular.js:25388$setViewValue @ angular.js:25360listener @ angular.js:21679eventHandler @ angular.js:3298
```

If I paste in a code before trying to type it in, it works, though.

callmehiphop commented:
I think I've address all of those now if you want to take another look. We now trigger `Element#select()` once that share code dialog pops up, so the user just needs to hit copy. (Think we should add a copy to clipboard button..?)

Good catch on all the form validation stuff, I fixed all that.

stephenplusplus commented:
@jgeewax -- we made a lot of progress on this, check out my fork http://stephenplusplus.github.io/phoenix/#/projects

We still need to work on some things but hopefully this is good enough for a quick demo. All feedback welcome!

stephenplusplus commented:
@callmehiphop is there anything new since the last couple commits you added on? Do we need to do anything before merging?

callmehiphop commented:
@stephenplusplus I haven't really touched this in about a week. There's still the outstanding firebase issue about the query syntax ( I think? ). I just put a quick fix in gh-pages.

stephenplusplus commented:
Added that and squashed into one. The rest of the To Dos can be tackled in future issues. Can you give this a look to make sure it meets the @callmehiphop standards?

callmehiphop commented on src/app/components/firebase/firebase.service.js:
```diff
@@ -17,7 +17,7 @@
 
       if (authData) {
-        userId = authData.uid;
+        $rootScope.USER_ID = userId = authData.uid;
```
Any reason why we're assigning this to `$rootScope`?

stephenplusplus replied:
Answered here: https://github.com/jgeewax/phoenix/pull/45#discussion_r45146451

callmehiphop commented on src/app/index.route.js:
```diff
@@ -15,4 +16,11 @@
+  /** @ngInject */
+  function runBlock($rootScope) {
+    $rootScope.$on('$stateChangeSuccess', function(e, toState) {
+      $rootScope.ACTIVE_STATE = toState;
```
1. See https://github.com/jgeewax/phoenix/pull/45/files#r45142055
2. Per the Angular community styleguide, we should only be putting one operation (run/config/etc.) per file, this would probably be better placed within https://github.com/jgeewax/phoenix/blob/master/src/app/index.run.js

stephenplusplus replied:
Answering for both ("why $rootScope?"), they were meant to be used in the view.

`USER_ID` didn't end up being used in a view, so it can be refactored to just return from an accessor method (`firebaseDriver.getUserId()`).

For `ACTIVE_STATE`, it's used here: https://github.com/stephenplusplus/phoenix/blob/e56f934ab3e54d058353f3a8565bc2e2c2f648f5/src/app/projects/projects.html#L3

stephenplusplus replied:
For point 2, that makes sense.

stephenplusplus replied:
Added a `getUserId()` for firebaseDriver.

callmehiphop commented on src/app/projects/projects.controller.js:
```diff
@@ -5,7 +6,183 @@
+    var projects = this;
+
+    projects.map = $projects.projects;
+    for (var projectId in projects.map) {
```
This is mostly a nitpick but, we could maybe clean this up with [`angular.forEach`](https://docs.angularjs.org/api/ng/function/angular.forEach)

stephenplusplus replied:
Done.

callmehiphop commented on src/app/projects/projects.html:
```diff
@@ -1,5 +1,82 @@
 <section layout="column" flex>
-  <navbar projects="$projects.projects" user="gapi.user"></navbar>
+  <navbar projects="projects.map" user="gapi.user"></navbar>
+  <div ng-if="ACTIVE_STATE.name === 'projects'">
```
Instead of an `ng-if` is there some way we could specify a different state or view for this?

stephenplusplus replied:
Would you mind leading that charge? I find this stuff quite complicated, and I never seem to strike the right solution. It'd probably just be a time waster for all of us to spend my time there :)

stephenplusplus replied:
Still need help here, sorry :(

callmehiphop replied:
Haven't forgot! Just been kinda holding off on this
